### ISSUE TITLE
non_zero_dists might be size 0

### ISSUE THREAD

**Issue by sleighsoft** [lmcinnes/umap#284]
This does not seem to be an issue with 

https://github.com/lmcinnes/umap/blob/d5d995625b5cfe55430771e4ec2f044533da7b4c/umap/umap_.py#L118

**Comment by lmcinnes** [comment]
I and a colleague on working on this issue. The problem really is to work out the best API for handling the occurrence of duplicates in the dataset. We will have a solution implemented soon, and that will impact this appropriately.

**Comment by sleighsoft** [comment]
With duplicates you mean nearest neighbors with the same distance to a point?
Like 10 points all being on the circumference of a point?

**Comment by lmcinnes** [comment]
No -- by that issue I meant, for example, having _x_ points that are all identical, and an ``n_neighbors`` of less than _x_. This also lines up with related issues that can occur if you have a large number of points all tied for the same distance away from a given point which can cause issue for NNDescent. Ideally at a high level we want to relatively cheaply test for such (pathological) cases and then either warn or provide an optional work-around. The solution to identical points is potentially to simply factor them out, embed the unique points, and then add the duplicates to the embedding as duplicates of the embedded points; this is the behaviour some users would like to see. The catch is that that is expensive to compute (dupe finding is hard in large datasets), potentially unnecessary for most cases, and has slightly surprising behaviour if you aren't expecting it. The goal was to add an option to do duplicate handling, and the ability to throw a warning and suggest you try duplicate handling  if potential issues are found. I have a colleague who was working on getting all of that done.

**Comment by jc-healy** [comment]
I've got the dense case finished and the few unit tests I've written seems
to be working well.  The next step is to get the sparse one finished (more
unit test) and a then the detection heuristics.

**Comment by sleighsoft** [comment]
Thank you for the clarification.
I am using the UMAP "topology/graph" idea for a slightly different purpose, that is why for me handling equidistant points was important.

**Comment by lmcinnes** [comment]
Are you able to substitute in a working solution for yourself in the meantime, or are you waiting for the issue to be fixed here?

**Comment by sleighsoft** [comment]
I have two concerns right now:
1. How do I handle k-NN in case of a single point having multiple equidistant furthest neighbors?
  - Especially when they are the "furthest" neighbors. So assume `k=4` and a point that has 4 equidistant nearest neighbors. Do I know broaden the radius to include a "second" neighbor?
2. How do I handle k-NN in case a point has e.g. a single 1-NN, 2-NN, 3-NN but then multiple 4-NN, e.g. when `k=4`.

**Comment by sleighsoft** [comment]
Doing it this way with a mask works
```python
@numba.njit(parallel=True)
def fast_knn_indices(X, n_neighbors):
    knn_indices = np.empty(X.shape, dtype=np.int64)
    mask = np.full_like(X, False, dtype=np.bool_)
    max_indices = np.empty((X.shape[0]), dtype=np.int64)
    for i in numba.prange(knn_indices.shape[0]):
        row = X[i]
        v = row.argsort(kind="quicksort")
        knn = row[v[n_neighbors - 1]]
        j = -1
        for j in range(n_neighbors, knn_indices.shape[1]):
            if row[v[j]] != knn:
                break
        knn_indices[i] = v
        mask[i, :j] = True
        max_indices[i] = j
    max_knn_index = max_indices.max()
    mask = mask[:, :max_knn_index]
    knn_indices = knn_indices[:, :max_knn_index]
    return knn_indices, mask
```

**Comment by sleighsoft** [comment]
I know I am slightly misusing this issue as a chat but Github doesn't really provide an open chat ...

@lmcinnes You might want to run `pip install tbb` or install `OpenMP` when you make use of `parallel=True` to avoid potential concurrency issues. See here: https://github.com/numba/numba/issues/4484

**Comment by lmcinnes** [comment]
Thanks -- I'll try to keep that in mind. You may want to target future changes against the 0.4dev branch since that is where most new functionality is headed. Do we need to add tbb to the requirements.txt? Does that seem a reasonable approach.

**Comment by sleighsoft** [comment]
Yeah, just remember tbb only works on 64 bit systems. So you will limit UMAP to 64bit (though that should not really be a problem in 2019...)

**Comment by sleighsoft** [comment]
@lmcinnes What do you think about supporting "dynamic" knn with multiple identical nearest neigbors as in part per above example of `parallel_knn_indices"?

Assume you have a point A and 3 points B|C|D are equidistant from point A. Currently if you specify `k=3` only point B and C will be considered neighbors but D is also one.

**Comment by sleighsoft** [comment]
Again misusing the issue as a chat.
You should close "stale" or "resolved" issues. It will grow out of control at some point.
If you want I'll close the ones that seem **done**. Just don't want to misuse my new repository rights.

**Comment by lmcinnes** [comment]
Fell free to close any that are fully resolved. You may want to add a comment to ones that are just stale as often other people are looking for the issue, and may add to it later.

**Comment by sleighsoft** [comment]
@lmcinnes Coming back to support for equidistant nearest neighbors. I have added a rough implementation and a small example in https://github.com/lmcinnes/umap/tree/equidistant_neighbors

When you run `devel/test.py` it will print the mean distance of the center point `[0,0]` to its neighbors which are all the same distance away from `[0,0]`. Running this on the `master` branch code will yield a higher on average distance of points around `[0,0]` after embedding with UMAP than it does on the `equidistant` branch.
